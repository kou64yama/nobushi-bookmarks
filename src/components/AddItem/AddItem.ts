import Vue from 'vue';
import Component from 'vue-class-component';
import { mapGetters, mapState } from 'vuex';

interface ValidationError extends Error {
  details: {
    path: string;
    message: string;
  }[];
}

@Component({
  computed: {
    ...mapGetters('items', ['getFieldErrors']),
    ...mapState('items', ['loading']),
  },
  watch: {
    open(this: AddItem) {
      this.name = '';
      this.url = '';
      this.description = '';
      this.$store.dispatch('items/clearError');
    },
  },
})
export default class AddItem extends Vue {
  public open = false;
  public name = '';
  public url = '';
  public description = '';

  public async submit() {
    const owner = this.$store.getters['auth/uid'];
    await this.$store.dispatch('items/add', {
      owner,
      name: this.name,
      url: this.url,
      description: this.description,
    });

    if (!this.$store.state.items.error) {
      this.open = false;
    }
  }
}
