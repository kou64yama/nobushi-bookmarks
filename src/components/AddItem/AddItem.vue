<template>
  <v-dialog width="480" v-model="open">
    <v-btn
      fab
      fixed
      bottom
      right
      :loading="loading"
      color="primary"
      slot="activator"
    >
      <v-icon>add</v-icon>
    </v-btn>
    <v-card>
      <v-card-text>
        <v-text-field
          label="Name"
          name="name"
          :error-messages="getFieldErrors('name')"
          v-model="name"
        />
        <v-text-field
          label="URL"
          name="url"
          :error-messages="getFieldErrors('url')"
          v-model="url"
        />
        <v-textarea
          label="Description"
          name="description"
          :error-messages="getFieldErrors('description')"
          v-model="description"
        />
      </v-card-text>
      <v-card-actions>
        <v-spacer/>
        <v-btn
          flat
          color="primary"
          :loading="loading"
          @click.native="open = false"
        >
          Cancel
        </v-btn>
        <v-btn
          depressed
          ref="add"
          color="primary"
          :loading="loading"
          @click.native="submit"
        >
          Add
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts" src="./AddItem.ts"></script>
