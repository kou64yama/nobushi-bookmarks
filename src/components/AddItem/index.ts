import AddItem from './AddItem.vue';

export default AddItem;
