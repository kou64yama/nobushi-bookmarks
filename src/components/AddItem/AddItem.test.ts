import Vuex, { Dispatch } from 'vuex';
import { shallowMount, createLocalVue, Stubs } from '@vue/test-utils';
import AddItem from './AddItem.vue';
import Vue from 'vue';

const localVue = createLocalVue();

localVue.use(Vuex);

const stubs: Stubs = {
  VDialog: true,
  VBtn: true,
  VIcon: true,
  VCard: true,
  VCardText: true,
  VTextField: true,
  VTextarea: true,
  VCardActions: true,
  VSpacer: true,
};

describe('AddItem', () => {
  const createStore = (dispatch: Dispatch = jest.fn()) =>
    Object.assign(
      new Vuex.Store({
        getters: {
          'auth/uid': () => 'test user',
        },
      }),
      {
        dispatch,
      },
    );

  it('should render correctly', () => {
    const store = createStore();
    const wrapper = shallowMount(AddItem, {
      localVue,
      store,
      stubs,
      computed: {
        loading: () => false,
        getFieldErrors: () => () => [],
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render correctly when loading', () => {
    const store = createStore();
    const wrapper = shallowMount(AddItem, {
      localVue,
      store,
      stubs,
      computed: {
        loading: () => true,
        getFieldErrors: () => () => [],
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should dispatch items/add when add button is clicked', () => {
    const store = createStore();
    const wrapper = shallowMount(AddItem, {
      localVue,
      store,
      stubs,
      computed: {
        loading: () => false,
        getFieldErrors: () => () => [],
      },
    });

    wrapper.vm.$data.name = 'Test Name';
    wrapper.vm.$data.url = 'https://example.com';
    wrapper.vm.$data.description = 'Test Description';
    wrapper.find({ ref: 'add' }).trigger('click');
    expect(store.dispatch).toHaveBeenLastCalledWith('items/add', {
      name: 'Test Name',
      url: 'https://example.com',
      description: 'Test Description',
      owner: 'test user',
    });
  });
});
