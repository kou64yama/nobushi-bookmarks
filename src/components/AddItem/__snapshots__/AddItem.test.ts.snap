// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItem should render correctly 1`] = `
<vdialog-stub
  width="480"
>
  <vbtn-stub
    bottom=""
    color="primary"
    fab=""
    fixed=""
    right=""
  >
    <vicon-stub>
      add
    </vicon-stub>
  </vbtn-stub>
   
  <vcard-stub>
    <vcardtext-stub>
      <vtextfield-stub
        error-messages=""
        label="Name"
        name="name"
      />
       
      <vtextfield-stub
        error-messages=""
        label="URL"
        name="url"
      />
       
      <vtextarea-stub
        error-messages=""
        label="Description"
        name="description"
      />
    </vcardtext-stub>
     
    <vcardactions-stub>
      <vspacer-stub />
       
      <vbtn-stub
        color="primary"
        flat=""
      >
        
        Cancel
      
      </vbtn-stub>
       
      <vbtn-stub
        color="primary"
        depressed=""
      >
        
        Add
      
      </vbtn-stub>
    </vcardactions-stub>
  </vcard-stub>
</vdialog-stub>
`;

exports[`AddItem should render correctly when loading 1`] = `
<vdialog-stub
  width="480"
>
  <vbtn-stub
    bottom=""
    color="primary"
    fab=""
    fixed=""
    loading="true"
    right=""
  >
    <vicon-stub>
      add
    </vicon-stub>
  </vbtn-stub>
   
  <vcard-stub>
    <vcardtext-stub>
      <vtextfield-stub
        error-messages=""
        label="Name"
        name="name"
      />
       
      <vtextfield-stub
        error-messages=""
        label="URL"
        name="url"
      />
       
      <vtextarea-stub
        error-messages=""
        label="Description"
        name="description"
      />
    </vcardtext-stub>
     
    <vcardactions-stub>
      <vspacer-stub />
       
      <vbtn-stub
        color="primary"
        flat=""
        loading="true"
      >
        
        Cancel
      
      </vbtn-stub>
       
      <vbtn-stub
        color="primary"
        depressed=""
        loading="true"
      >
        
        Add
      
      </vbtn-stub>
    </vcardactions-stub>
  </vcard-stub>
</vdialog-stub>
`;
